File: ios/App/Sources/Features/Settings/UserMemoryStore.swift
import Foundation
import Combine

/// Structured user memory (Settings → Memory), private on device.
/// Mirrors desktop `UserMemoryStore`: You / Topics / Areas, freeform add,
/// detail edit, and import paste from another AI provider.
@MainActor
final class UserMemoryStore: ObservableObject {
    static let shared = UserMemoryStore()

    enum Category: String, Codable, CaseIterable, Identifiable {
        case you
        case topic
        case area

        var id: String { rawValue }

        var label: String {
            switch self {
            case .you: return "You"
            case .topic: return "Topics"
            case .area: return "Areas"
            }
        }

        static let displayOrder: [Category] = [.you, .topic, .area]
    }

    struct Entry: Identifiable, Codable, Equatable, Hashable {
        var id: String
        var category: Category
        var title: String
        var summary: String
        var details: [String]
        var updatedAt: Date
    }

    /// One row in the activity log: a single mutation the user can undo.
    /// The `before` snapshot lets us restore a prior `details` / `summary`
    /// when the user undoes an `add` or `update`; `new` lets us delete the
    /// entry that was just created. `source` distinguishes auto-save (from
    /// chat) from explicit edits (manage screen / detail sheet).
    struct ActivityEntry: Identifiable, Codable, Equatable, Hashable {
        var id: String
        var timestamp: Date
        var kind: Kind
        var entryID: String
        var entryTitle: String
        var detailPreview: String
        var before: Entry?
        var after: Entry?
        var source: Source

        enum Kind: String, Codable {
            case add
            case update
            case forget
            case rename
        }

        enum Source: String, Codable {
            case chatAutoSave = "chat"
            case userEdit = "user"
        }
    }

    @Published private(set) var entries: [Entry] = []
    @Published private(set) var activity: [ActivityEntry] = []

    private let key = "userMemory.entries.v1"
    private let activityKey = "userMemory.activity.v1"
    private let activityMaxAge: TimeInterval = 60 * 60 * 24 * 30 // 30 days
    private let activityMaxCount = 200

    /// Prompt users copy into another AI product when importing memory.
    static let importPrompt = """
    Export all of my stored memories and any context you've learned about me from past conversations. Preserve my words verbatim where possible, especially for instructions and preferences.

    ## Categories (output in this order):

    ### You — Profile
    - Summary line
    - Bullet details (name, role, location, preferences)

    ### Topics
    One section per interest/topic with a short summary and bullets.

    ### Areas
    One section per project/product/domain with a short summary and bullets.
    """

    private init() {
        load()
        pruneActivity()
    }

    var isEmpty: Bool { entries.isEmpty }

    func list() -> [Entry] {
        entries.sorted { $0.updatedAt > $1.updatedAt }
    }

    func byCategory(_ category: Category) -> [Entry] {
        list().filter { $0.category == category }
    }

    func entry(id: String) -> Entry? {
        entries.first { $0.id == id }
    }

    /// Recent activity rows, newest first. Caller can filter by source.
    func activityList(source: ActivityEntry.Source? = nil, limit: Int = 50) -> [ActivityEntry] {
        let pool = source.map { s in activity.filter { $0.source == s } } ?? activity
        return Array(pool.sorted { $0.timestamp > $1.timestamp }.prefix(limit))
    }

    @discardableResult
    func upsert(
        id: String? = nil,
        category: Category,
        title: String,
        summary: String,
        details: [String]
    ) -> Entry {
        let now = Date()
        if let id, let idx = entries.firstIndex(where: { $0.id == id }) {
            entries[idx].category = category
            entries[idx].title = title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                ? entries[idx].title
                : title.trimmingCharacters(in: .whitespacesAndNewlines)
            entries[idx].summary = summary.trimmingCharacters(in: .whitespacesAndNewlines)
            entries[idx].details = details
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            entries[idx].updatedAt = now
            save()
            return entries[idx]
        }
        let entry = Entry(
            id: id ?? "mem_\(UUID().uuidString.prefix(8).lowercased())",
            category: category,
            title: title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                ? "Untitled"
                : title.trimmingCharacters(in: .whitespacesAndNewlines),
            summary: summary.trimmingCharacters(in: .whitespacesAndNewlines),
            details: details
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty },
            updatedAt: now
        )
        entries.insert(entry, at: 0)
        save()
        return entry
    }

    func delete(id: String) {
        entries.removeAll { $0.id == id }
        save()
    }

    @discardableResult
    func addFreeformFact(_ text: String) -> Entry? {
        let fact = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !fact.isEmpty else { return nil }
        let bullet = fact.prefix(1).uppercased() + fact.dropFirst()
        if var profile = byCategory(.you).first(where: { $0.title.lowercased() == "profile" }) {
            if !profile.details.contains(where: { $0.caseInsensitiveCompare(String(bullet)) == .orderedSame }) {
                profile.details.append(String(bullet))
            }
            if profile.summary.isEmpty {
                profile.summary = String(bullet.prefix(80))
            }
            profile.updatedAt = Date()
            if let idx = entries.firstIndex(where: { $0.id == profile.id }) {
                entries[idx] = profile
            }
            save()
            return profile
        }
        return upsert(
            category: .you,
            title: "Profile",
            summary: String(bullet.prefix(80)),
            details: [String(bullet)]
        )
    }

    @discardableResult
    func applyEntryCommand(id: String, command: String) -> Entry? {
        guard let entry = entry(id: id) else { return nil }
        let cmd = command.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cmd.isEmpty else { return entry }
        let kind = classifyCommand(cmd)
        let updated = mutateEntry(entry: entry, kind: kind, command: cmd)
        guard let updated, updated != entry else { return entry }
        return commitMutation(
            kind: activityKind(for: kind),
            before: entry,
            after: updated,
            source: .userEdit
        )
    }

    /// Apply a structured action parsed from a chat reply. Returns the
    /// resulting entry, or nil if the action was a no-op.
    @discardableResult
    func applyAction(_ action: ParsedAction) -> Entry? {
        switch action {
        case let .add(category, title, summary, details):
            // Match by title within the same category; create if missing.
            let target = entry(forTitle: title, category: category)
            if var target {
                let before = target
                let newDetails = details
                    .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                    .filter { !$0.isEmpty }
                let combined = mergeDetails(existing: target.details, adding: newDetails)
                if combined == target.details && (summary.isEmpty || summary == target.summary) {
                    return target
                }
                target.details = combined
                if !summary.isEmpty { target.summary = summary.trimmingCharacters(in: .whitespacesAndNewlines) }
                target.updatedAt = Date()
                return commitMutation(
                    kind: .update,
                    before: before,
                    after: target,
                    source: .chatAutoSave
                )
            } else {
                let created = upsert(
                    category: category,
                    title: title,
                    summary: summary,
                    details: details
                )
                recordActivity(kind: .add, before: nil, after: created, source: .chatAutoSave)
                return created
            }
        case let .forget(target):
            guard let victim = entry(forTitleOrContains: target) else { return nil }
            let before = victim
            let next = victim.details.filter { !$0.localizedCaseInsensitiveContains(target) }
            if next.count == victim.details.count { return nil }
            var updated = victim
            updated.details = next
            updated.updatedAt = Date()
            return commitMutation(
                kind: .forget,
                before: before,
                after: updated,
                source: .chatAutoSave
            )
        case let .rename(target, value):
            guard let victim = entry(forTitleOrContains: target) else { return nil }
            let before = victim
            var updated = victim
            updated.title = value
            updated.updatedAt = Date()
            return commitMutation(
                kind: .rename,
                before: before,
                after: updated,
                source: .chatAutoSave
            )
        case let .setSummary(target, value):
            guard let victim = entry(forTitleOrContains: target) else { return nil }
            let before = victim
            var updated = victim
            updated.summary = value
            updated.updatedAt = Date()
            return commitMutation(
                kind: .update,
                before: before,
                after: updated,
                source: .chatAutoSave
            )
        case let .setDetails(target, value):
            guard let victim = entry(forTitleOrContains: target) else { return nil }
            let before = victim
            var updated = victim
            updated.details = value
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            updated.updatedAt = Date()
            return commitMutation(
                kind: .update,
                before: before,
                after: updated,
                source: .chatAutoSave
            )
        }
    }

    /// Structured action emitted by the chat parser. Mirrors the desktop
    /// `MemoryAction` union minus the wire-format-specific details.
    enum ParsedAction: Equatable {
        case add(category: Category, title: String, summary: String, details: [String])
        case forget(target: String)
        case rename(target: String, value: String)
        case setSummary(target: String, value: String)
        case setDetails(target: String, value: [String])
    }

    // MARK: - Private mutation helpers

    private enum CommandKind {
        case forget(topic: String)
        case setSummary(value: String)
        case rename(value: String)
        case appendOrReplaceFact(fact: String)
    }

    private func classifyCommand(_ cmd: String) -> CommandKind {
        let lower = cmd.lowercased()
        if lower.hasPrefix("forget ") || lower.hasPrefix("remove ") || lower.hasPrefix("delete ") {
            let topic = cmd
                .replacingOccurrences(of: #"^(forget|remove|delete)\s+"#, with: "", options: .regularExpression)
                .trimmingCharacters(in: .whitespacesAndNewlines)
            return .forget(topic: topic)
        }
        if let range = lower.range(of: #"^(change|set) summary to\s+"#, options: .regularExpression) {
            let drop = cmd.distance(from: cmd.startIndex, to: range.upperBound)
            return .setSummary(value: String(cmd.dropFirst(drop)).trimmingCharacters(in: .whitespacesAndNewlines))
        }
        if lower.hasPrefix("rename to ") {
            let name = String(cmd.dropFirst("rename to ".count)).trimmingCharacters(in: .whitespacesAndNewlines)
            return .rename(value: name)
        }
        var fact = cmd
        if let r = fact.range(of: #"^remember that(?: i)?\s+"#, options: [.regularExpression, .caseInsensitive]) {
            fact = String(fact[r.upperBound...])
        } else if fact.lowercased().hasPrefix("remember ") {
            fact = String(fact.dropFirst("remember ".count))
        } else if fact.lowercased().hasPrefix("add ") {
            fact = String(fact.dropFirst("add ".count))
        }
        return .appendOrReplaceFact(fact: fact.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    private func mutateEntry(entry: Entry, kind: CommandKind, command: String) -> Entry? {
        var entry = entry
        switch kind {
        case let .forget(topic):
            guard !topic.isEmpty else { return nil }
            let next = entry.details.filter { !$0.localizedCaseInsensitiveContains(topic) }
            if next.count == entry.details.count { return nil }
            entry.details = next
            if entry.summary.localizedCaseInsensitiveContains(topic) {
                entry.summary = next.first ?? entry.summary
            }
        case let .setSummary(value):
            entry.summary = value
        case let .rename(value):
            guard !value.isEmpty else { return nil }
            entry.title = value
        case let .appendOrReplaceFact(fact):
            guard !fact.isEmpty else { return nil }
            let bullet = fact.prefix(1).uppercased() + fact.dropFirst()
            if entry.details.count <= 1 {
                entry.details = [String(bullet)]
                if entry.summary.isEmpty
                    || entry.summary.caseInsensitiveCompare(entry.details.first ?? "") != .orderedSame {
                    entry.summary = String(bullet.prefix(80))
                }
            } else if !entry.details.contains(where: { $0.caseInsensitiveCompare(String(bullet)) == .orderedSame }) {
                entry.details.append(String(bullet))
            } else {
                return nil // duplicate fact
            }
        }
        entry.updatedAt = Date()
        return entry
    }

    private func activityKind(for kind: CommandKind) -> ActivityEntry.Kind {
        switch kind {
        case .forget: return .forget
        case .rename: return .rename
        case .setSummary, .appendOrReplaceFact: return .update
        }
    }

    private func commitMutation(
        kind: ActivityEntry.Kind,
        before: Entry,
        after: Entry,
        source: ActivityEntry.Source
    ) -> Entry {
        if let idx = entries.firstIndex(where: { $0.id == after.id }) {
            entries[idx] = after
        } else {
            entries.insert(after, at: 0)
        }
        save()
        recordActivity(kind: kind, before: before, after: after, source: source)
        return after
    }

    private func entry(forTitle title: String, category: Category) -> Entry? {
        let lower = title.lowercased()
        return entries.first { $0.category == category && $0.title.lowercased() == lower }
    }

    private func entry(forTitleOrContains needle: String) -> Entry? {
        let lower = needle.lowercased()
        if let exact = entries.first(where: { $0.title.lowercased() == lower }) {
            return exact
        }
        return entries.first { $0.details.contains(where: { $0.lowercased().contains(lower) }) }
            ?? entries.first { $0.summary.lowercased().contains(lower) }
    }

    private func mergeDetails(existing: [String], adding: [String]) -> [String] {
        var out = existing
        for d in adding {
            if !out.contains(where: { $0.caseInsensitiveCompare(d) == .orderedSame }) {
                out.append(d)
            }
        }
        return out
    }

    @discardableResult
    func importFromText(_ raw: String) -> Int {
        let text = raw
            .replacingOccurrences(of: "\r\n", with: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return 0 }

        let blocks = Self.splitImportBlocks(text)
        if blocks.isEmpty {
            let lines = text
                .components(separatedBy: .newlines)
                .map { $0.replacingOccurrences(of: #"^[-*•]\s+"#, with: "", options: .regularExpression) }
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            upsert(
                category: .you,
                title: "Imported",
                summary: String(text.prefix(100)).replacingOccurrences(of: "\n", with: " "),
                details: Array(lines.prefix(40))
            )
            return 1
        }
        for block in blocks {
            upsert(
                category: block.category,
                title: block.title,
                summary: block.summary,
                details: block.details
            )
        }
        return blocks.count
    }

    /// System prompt blob for chat injection.
    func formatForSystem() -> String {
        var parts: [String] = []
        for cat in Category.displayOrder {
            let items = byCategory(cat)
            guard !items.isEmpty else { continue }
            parts.append("## \(cat.label)")
            for e in items {
                parts.append("### \(e.title)")
                if !e.summary.isEmpty { parts.append(e.summary) }
                for d in e.details { parts.append("- \(d)") }
            }
        }
        return parts.joined(separator: "\n")
    }

    // MARK: - Persistence

    private func load() {
        if let data = UserDefaults.standard.data(forKey: key),
           let decoded = try? JSONDecoder().decode([Entry].self, from: data) {
            entries = decoded
        }
        if let data = UserDefaults.standard.data(forKey: activityKey),
           let decoded = try? JSONDecoder().decode([ActivityEntry].self, from: data) {
            activity = decoded
        }
    }

    private func save() {
        if let data = try? JSONEncoder().encode(entries) {
            UserDefaults.standard.set(data, forKey: key)
        }
        objectWillChange.send()
    }

    private func saveActivity() {
        if let data = try? JSONEncoder().encode(activity) {
            UserDefaults.standard.set(data, forKey: activityKey)
        }
    }

    private func pruneActivity() {
        let cutoff = Date().addingTimeInterval(-activityMaxAge)
        let kept = activity.filter { $0.timestamp >= cutoff }
        let trimmed = kept.count > activityMaxCount
            ? Array(kept.suffix(activityMaxCount))
            : kept
        if trimmed.count != activity.count {
            activity = trimmed
            saveActivity()
        }
    }

    // MARK: - Activity log

    /// Record a mutation so the user can see and undo it. `source` is
    /// `.userEdit` for explicit manage-screen edits and `.chatAutoSave` for
    /// tags parsed from a chat reply.
    func recordActivity(
        kind: ActivityEntry.Kind,
        before: Entry?,
        after: Entry?,
        source: ActivityEntry.Source
    ) {
        let id = "act_\(UUID().uuidString.prefix(8).lowercased())"
        let previewEntry = after ?? before
        let preview = previewEntry?.summary.isEmpty == false
            ? previewEntry!.summary
            : (previewEntry?.details.first ?? previewEntry?.title ?? "")
        let entry = ActivityEntry(
            id: id,
            timestamp: Date(),
            kind: kind,
            entryID: after?.id ?? before?.id ?? "",
            entryTitle: previewEntry?.title ?? "",
            detailPreview: preview,
            before: before,
            after: after,
            source: source
        )
        activity.append(entry)
        if activity.count > activityMaxCount {
            activity = Array(activity.suffix(activityMaxCount))
        }
        saveActivity()
    }

    /// Undo a single activity row. Returns true on success. For every kind,
    /// restoring the entry means putting back the `before` snapshot. For
    /// `add` (where there is no `before`), the entry is removed entirely.
    @discardableResult
    func undoActivity(id: String) -> Bool {
        guard let idx = activity.firstIndex(where: { $0.id == id }) else { return false }
        let row = activity[idx]
        if let before = row.before {
            if let eidx = entries.firstIndex(where: { $0.id == before.id }) {
                entries[eidx] = before
            } else {
                entries.insert(before, at: 0)
            }
        } else {
            entries.removeAll { $0.id == row.entryID }
        }
        activity.remove(at: idx)
        save()
        saveActivity()
        return true
    }

    // MARK: - Import parse

    private struct ImportBlock {
        var category: Category
        var title: String
        var summary: String
        var details: [String]
    }

    private static func splitImportBlocks(_ text: String) -> [ImportBlock] {
        let lines = text.components(separatedBy: .newlines)
        var blocks: [ImportBlock] = []
        var category: Category = .you
        var title = ""
        var summary = ""
        var details: [String] = []
        var sawHeading = false

        func flush() {
            guard !title.isEmpty || !details.isEmpty || !summary.isEmpty else { return }
            blocks.append(ImportBlock(
                category: category,
                title: title.isEmpty ? (category == .you ? "Profile" : "Untitled") : title,
                summary: summary.isEmpty ? (details.first ?? "") : summary,
                details: details.isEmpty ? (summary.isEmpty ? [] : [summary]) : details
            ))
            title = ""
            summary = ""
            details = []
        }

        for rawLine in lines {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !line.isEmpty else { continue }

            if let catMatch = line.range(
                of: #"^(?:#{1,3}\s*)?(You|Topics?|Areas?)\b(?:\s*[—–:-]\s*(.+))?$"#,
                options: [.regularExpression, .caseInsensitive]
            ) {
                flush()
                sawHeading = true
                let matched = String(line[catMatch])
                let label = matched.lowercased()
                if label.contains("topic") { category = .topic }
                else if label.contains("area") { category = .area }
                else { category = .you }

                if let restRange = matched.range(of: #"[—–:-]\s*(.+)$"#, options: .regularExpression) {
                    var rest = String(matched[restRange])
                    rest = rest.replacingOccurrences(of: #"^[—–:-]\s*"#, with: "", options: .regularExpression)
                    if !rest.isEmpty { title = rest }
                    else if category == .you { title = "Profile" }
                } else if category == .you {
                    title = "Profile"
                }
                continue
            }

            if let h = line.range(of: #"^#{1,3}\s+(.+)$"#, options: .regularExpression) {
                flush()
                sawHeading = true
                title = String(line[h])
                    .replacingOccurrences(of: #"^#{1,3}\s+"#, with: "", options: .regularExpression)
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                if category == .you && !title.localizedCaseInsensitiveContains("profile") {
                    if title.range(of: #"app|project|product|startup|company|kind|work"#, options: [.regularExpression, .caseInsensitive]) != nil {
                        category = .area
                    } else if title.range(of: #"name|role|location|bio"#, options: [.regularExpression, .caseInsensitive]) == nil {
                        category = .topic
                    }
                }
                continue
            }

            let bullet = line
                .replacingOccurrences(of: #"^[-*•]\s+"#, with: "", options: .regularExpression)
                .replacingOccurrences(of: #"^\d+\.\s+"#, with: "", options: .regularExpression)
            if bullet != line || line.hasPrefix("-") || line.hasPrefix("*") || line.hasPrefix("•") {
                details.append(bullet.trimmingCharacters(in: .whitespacesAndNewlines))
                continue
            }

            if summary.isEmpty {
                summary = line
            } else {
                details.append(line)
            }
        }
        flush()
        return sawHeading ? blocks : []
    }
}

extension UserMemoryStore.Entry {
    var relativeUpdated: String {
        let diff = Date().timeIntervalSince(updatedAt)
        if diff < 60 { return "just now" }
        if diff < 3600 { return "\(Int(diff / 60))m ago" }
        if diff < 172_800 { return "\(Int(diff / 3600))h ago" }
        if diff < 1_209_600 { return "\(Int(diff / 86_400))d ago" }
        let f = DateFormatter()
        f.dateStyle = .medium
        f.timeStyle = .none
        return f.string(from: updatedAt)
    }
}

File: desktop-server/src/client/user-memory-store.ts
/**
 * User (profile) memory — structured entries for Settings → Memory.
 *
 * Mirrors Claude-style memory: You / Topics / Areas, freeform add/edit,
 * detail view, and import paste from another AI provider.
 * Stored in localStorage only (device-private).
 */
import type { StorageLike } from './history-store.js';

const KEY = 'apc.userMemory.v1';

export type MemoryCategory = 'you' | 'topic' | 'area';

export interface MemoryEntry {
  id: string;
  category: MemoryCategory;
  /** Short label, e.g. "Profile", "Gaming", "Kind365". */
  title: string;
  /** One-line blurb shown in the list. */
  summary: string;
  /** Bullet details shown on the detail screen. */
  details: string[];
  updatedAt: number;
}

export interface UserMemoryState {
  entries: MemoryEntry[];
}

export const MEMORY_CATEGORY_LABELS: Record<MemoryCategory, string> = {
  you: 'You',
  topic: 'Topics',
  area: 'Areas',
};

export const MEMORY_CATEGORY_ORDER: MemoryCategory[] = ['you', 'topic', 'area'];

/** Prompt users copy into another AI product when importing memory. */
export const MEMORY_IMPORT_PROMPT = `Export all of my stored memories and any context you've learned about me from past conversations. Preserve my words verbatim where possible, especially for instructions and preferences.

## Categories (output in this order):

### You — Profile
- Summary line
- Bullet details (name, role, location, preferences)

### Topics
One section per interest/topic with a short summary and bullets.

### Areas
One section per project/product/domain with a short summary and bullets.`;

function uid(): string {
  return `mem_${Math.random().toString(36).slice(2, 10)}`;
}

function normalizeEntry(raw: Partial<MemoryEntry> & { id: string; title: string }): MemoryEntry {
  const cat = raw.category;
  const category: MemoryCategory = cat === 'topic' || cat === 'area' || cat === 'you' ? cat : 'you';
  return {
    id: raw.id,
    category,
    title: (raw.title || 'Untitled').trim() || 'Untitled',
    summary: (raw.summary ?? '').trim(),
    details: Array.isArray(raw.details)
      ? raw.details.map((d) => String(d).trim()).filter(Boolean)
      : [],
    updatedAt: typeof raw.updatedAt === 'number' ? raw.updatedAt : Date.now(),
  };
}

export class UserMemoryStore {
  private entries: MemoryEntry[] = [];

  constructor(private storage: StorageLike) {
    this.load();
  }

  load(): void {
    try {
      const raw = this.storage.getItem(KEY);
      if (!raw) {
        this.entries = [];
        return;
      }
      const parsed = JSON.parse(raw) as { entries?: Partial<MemoryEntry>[] };
      this.entries = Array.isArray(parsed.entries)
        ? parsed.entries
            .filter(
              (e): e is Partial<MemoryEntry> & { id: string; title: string } =>
                !!e && typeof e.id === 'string' && typeof e.title === 'string'
            )
            .map(normalizeEntry)
        : [];
    } catch {
      this.entries = [];
    }
  }

  persist(): void {
    this.storage.setItem(KEY, JSON.stringify({ entries: this.entries }));
  }

  list(): MemoryEntry[] {
    return [...this.entries].sort((a, b) => b.updatedAt - a.updatedAt);
  }

  get(id: string): MemoryEntry | undefined {
    return this.entries.find((e) => e.id === id);
  }

  byCategory(category: MemoryCategory): MemoryEntry[] {
    return this.list().filter((e) => e.category === category);
  }

  upsert(entry: Omit<MemoryEntry, 'id' | 'updatedAt'> & { id?: string }): MemoryEntry {
    const now = Date.now();
    if (entry.id) {
      const existing = this.get(entry.id);
      if (existing) {
        existing.category = entry.category;
        existing.title = entry.title.trim() || existing.title;
        existing.summary = entry.summary.trim();
        existing.details = entry.details.map((d) => d.trim()).filter(Boolean);
        existing.updatedAt = now;
        this.persist();
        return existing;
      }
    }
    const created = normalizeEntry({
      id: entry.id ?? uid(),
      category: entry.category,
      title: entry.title,
      summary: entry.summary,
      details: entry.details,
      updatedAt: now,
    });
    this.entries.unshift(created);
    this.persist();
    return created;
  }

  delete(id: string): void {
    this.entries = this.entries.filter((e) => e.id !== id);
    this.persist();
  }

  /**
   * Freeform fact at the Memory root input (e.g. "My dog's name is Beans").
   * Appends to Profile under You, creating it when needed.
   */
  addFreeformFact(text: string): MemoryEntry | null {
    const fact = text.trim();
    if (!fact) return null;
    const bullet = fact.charAt(0).toUpperCase() + fact.slice(1);
    let profile = this.byCategory('you').find((e) => e.title.toLowerCase() === 'profile');
    if (!profile) {
      profile = this.upsert({
        category: 'you',
        title: 'Profile',
        summary: bullet.length > 80 ? `${bullet.slice(0, 77)}…` : bullet,
        details: [bullet],
      });
      return profile;
    }
    if (!profile.details.some((d) => d.toLowerCase() === bullet.toLowerCase())) {
      profile.details = [...profile.details, bullet];
    }
    if (!profile.summary.trim()) {
      profile.summary = bullet.length > 80 ? `${bullet.slice(0, 77)}…` : bullet;
    }
    profile.updatedAt = Date.now();
    this.persist();
    return profile;
  }

  /**
   * Natural-language edit for a single entry (detail screen).
   * "forget X" / "remove X" drops matching bullets; otherwise appends a note.
   */
  applyEntryCommand(id: string, command: string): MemoryEntry | null {
    const entry = this.get(id);
    if (!entry) return null;
    const cmd = command.trim();
    if (!cmd) return entry;
    const lower = cmd.toLowerCase();

    if (lower.startsWith('forget ') || lower.startsWith('remove ') || lower.startsWith('delete ')) {
      const topic = cmd.replace(/^(forget|remove|delete)\s+/i, '').trim();
      if (topic) {
        const nextDetails = entry.details.filter(
          (d) => !d.toLowerCase().includes(topic.toLowerCase())
        );
        if (nextDetails.length !== entry.details.length) {
          entry.details = nextDetails;
        } else {
          entry.details = [...entry.details, `Note: user asked to forget “${topic}”.`];
        }
        if (entry.summary.toLowerCase().includes(topic.toLowerCase())) {
          entry.summary = nextDetails[0] ?? entry.summary;
        }
      }
    } else if (/^change summary to\s+/i.test(cmd) || /^set summary to\s+/i.test(cmd)) {
      entry.summary = cmd.replace(/^(change|set) summary to\s+/i, '').trim();
    } else if (/^rename to\s+/i.test(cmd)) {
      entry.title = cmd.replace(/^rename to\s+/i, '').trim() || entry.title;
    } else {
      const fact = cmd
        .replace(/^remember that(?: i)?\s+/i, '')
        .replace(/^remember\s+/i, '')
        .replace(/^add\s+/i, '')
        .trim();
      const bullet = fact.charAt(0).toUpperCase() + fact.slice(1);
      if (bullet && !entry.details.some((d) => d.toLowerCase() === bullet.toLowerCase())) {
        entry.details = [...entry.details, bullet];
      }
    }
    entry.updatedAt = Date.now();
    this.persist();
    return entry;
  }

  /**
   * Parse pasted export from another AI into structured entries.
   * Best-effort markdown headings; falls back to one Imported note.
   */
  importFromText(raw: string): number {
    const text = raw.replace(/\r\n/g, '\n').trim();
    if (!text) return 0;

    const blocks = splitImportBlocks(text);
    if (blocks.length === 0) {
      this.upsert({
        category: 'you',
        title: 'Imported',
        summary: text.slice(0, 100).replace(/\n/g, ' '),
        details: text
          .split(/\n+/)
          .map((l) => l.replace(/^[-*•]\s*/, '').trim())
          .filter(Boolean)
          .slice(0, 40),
      });
      return 1;
    }

    let count = 0;
    for (const block of blocks) {
      this.upsert({
        category: block.category,
        title: block.title,
        summary: block.summary,
        details: block.details,
      });
      count += 1;
    }
    return count;
  }

  /** System prompt blob for chat injection. */
  formatForSystem(): string {
    const parts: string[] = [];
    for (const cat of MEMORY_CATEGORY_ORDER) {
      const items = this.byCategory(cat);
      if (items.length === 0) continue;
      parts.push(`## ${MEMORY_CATEGORY_LABELS[cat]}`);
      for (const e of items) {
        parts.push(`### ${e.title}`);
        if (e.summary) parts.push(e.summary);
        for (const d of e.details) {
          parts.push(`- ${d}`);
        }
      }
    }
    return parts.join('\n');
  }

  isEmpty(): boolean {
    return this.entries.length === 0;
  }
}

interface ImportBlock {
  category: MemoryCategory;
  title: string;
  summary: string;
  details: string[];
}

function splitImportBlocks(text: string): ImportBlock[] {
  const lines = text.split('\n');
  const blocks: ImportBlock[] = [];
  let category: MemoryCategory = 'you';
  let title = '';
  let summary = '';
  let details: string[] = [];
  let sawHeading = false;

  const flush = () => {
    if (!title && details.length === 0 && !summary) return;
    blocks.push({
      category,
      title: title || (category === 'you' ? 'Profile' : 'Untitled'),
      summary: summary || details[0] || '',
      details: details.length ? details : summary ? [summary] : [],
    });
    title = '';
    summary = '';
    details = [];
  };

  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (!line) continue;

    // Category headers: "## You", "### Topics", "You — Profile", etc.
    const catMatch = line.match(/^(?:#{1,3}\s*)?(You|Topics?|Areas?)\b(?:\s*[—–:-]\s*(.+))?$/i);
    if (catMatch) {
      flush();
      sawHeading = true;
      const label = catMatch[1]!.toLowerCase();
      if (label.startsWith('topic')) category = 'topic';
      else if (label.startsWith('area')) category = 'area';
      else category = 'you';
      const rest = (catMatch[2] ?? '').trim();
      if (rest) {
        title = rest;
      } else if (category === 'you') {
        title = 'Profile';
      }
      continue;
    }

    // Generic ### Title heading
    const hMatch = line.match(/^#{1,3}\s+(.+)$/);
    if (hMatch) {
      flush();
      sawHeading = true;
      title = hMatch[1]!.trim();
      // Infer category from keywords if still default
      if (category === 'you' && !/profile|about me|who i am/i.test(title)) {
        if (/app|project|product|startup|company|kind|work/i.test(title)) {
          category = 'area';
        } else if (!/name|role|location|bio/i.test(title)) {
          category = 'topic';
        }
      }
      continue;
    }

    const bullet = line.replace(/^[-*•]\s+/, '').replace(/^\d+\.\s+/, '');
    if (bullet !== line || line.startsWith('-') || line.startsWith('*') || line.startsWith('•')) {
      details.push(bullet.trim());
      continue;
    }

    if (!summary) {
      summary = line;
    } else {
      details.push(line);
    }
  }

  flush();
  return sawHeading ? blocks : [];
}

export function relativeMemoryTime(ts: number, now = Date.now()): string {
  const diff = Math.max(0, now - ts);
  const min = Math.floor(diff / 60_000);
  if (min < 1) return 'just now';
  if (min < 60) return `${min}m ago`;
  const hr = Math.floor(min / 60);
  if (hr < 48) return `${hr}h ago`;
  const day = Math.floor(hr / 24);
  if (day < 14) return `${day}d ago`;
  return new Date(ts).toLocaleDateString(undefined, {
    month: 'short',
    day: 'numeric',
  });
}
